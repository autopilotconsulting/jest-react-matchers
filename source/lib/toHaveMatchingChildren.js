import { describeReactElement } from 'describe-react-element';
import ReactMatcherUtils from './ReactMatcherUtils';
import groupArray from './groupArray';

import flatten from 'array-flatten';

function toHaveMatchingChildren(actualElement, ...expectedElements) {
  const actualDescription = describeReactElement(actualElement, 1);
  const children = actualElement.props.children;

  if (typeof(children) === 'undefined') { 
    const message = [
      'Actual element has no children:',
      actualDescription
    ].join("\n");

    return { pass: false, message };
  }

  expectedElements = flatten([expectedElements]);
  const { true: hits, false: misses } = groupArray(expectedElements, (element) => {
    return ReactMatcherUtils.collectionHasMatchForElement(this.equals, children, element);
  });
  
  const isCorrect = this.isNot ? !hits : !misses;
  const result = { pass: isCorrect && !this.isNot };

  if (isCorrect) { return result; }

  const describeExpected = (element) => describeReactElement(element, 1);
  if (this.isNot) {
    const hitDescriptions = hits.map(describeExpected);

    result.message = [
      'Expected not to find children:',
      ...hitDescriptions,
      '',
      'in:',
      actualDescription,
    ].join("\n");
  } else {
    const missDescriptions = misses.map(describeExpected);

    result.message = [
      'Expected to find children:',
      ...missDescriptions,
      '',
      'in:',
      actualDescription,
    ].join("\n");
  }

  return result;
};

export default toHaveMatchingChildren;