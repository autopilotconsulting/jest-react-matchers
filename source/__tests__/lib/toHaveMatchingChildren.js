import React from 'react';
import { equals } from '../../../support/jasmine_utils';

import toHaveMatchingChildren from '../../lib/toHaveMatchingChildren.js';

describe('ReactMatchers', () => {
  describe('#toHaveMatchingChildren', () => {
    let matcherContext;
    
    const actual = (
      <div>
        <li>Pippy</li>
        <li>Hatch</li>
        <li data-tis-wife={true}>Lu</li>
      </div>
    );

    beforeEach(() => {
      matcherContext = { equals, isNot: false };
    });

    it('should pass if all expected children are present', () => {
      let result = toHaveMatchingChildren.call(matcherContext, actual, <li />);
      expect(result.pass).toBe(true);

      result = toHaveMatchingChildren.call(matcherContext, actual, [<li>Pippy</li>, <li>Hatch</li>]);
      expect(result.pass).toBe(true);

      result = toHaveMatchingChildren.call(matcherContext, actual, <li>Pippy</li>, <li>Hatch</li>);
      expect(result.pass).toBe(true);
    });

    it('should not pass if any expected children are missing', () => {
      let result = toHaveMatchingChildren.call(matcherContext, actual, [<li>Pippy</li>, <span />]);
      expect(result.pass).toBe(false);
    });

    describe('when negated', () => {
      it('should not "pass" if all unexpected elements are missing', () => {
        matcherContext.isNot = true;
        let result = toHaveMatchingChildren.call(matcherContext, actual, [<marquee />, <span />]);
        expect(result.pass).toBe(false);
      });

      it('should "pass" if any unexpected children are present', () => {
        matcherContext.isNot = true;
        let result = toHaveMatchingChildren.call(matcherContext, actual, [<li>Pippy</li>, <span />]);
        expect(result.pass).toBe(true);
      });
    })

    it('should fail if the actual element has no children', () => {
      let actual = <div />;

      const result = toHaveMatchingChildren.call(matcherContext, actual, <h1 />);
      expect(result.pass).toBe(false);
      expect(result.message).toMatchSnapshot();
    });

    it('should describe a failure appropriately', () => {
      const expected = [
        <li data-tis-wife={true} />,
        <li>Lu</li>,
        <div data-awesome={true}>An awesome div!!</div>,
        <span>Lame Span</span>
      ];

      const result = toHaveMatchingChildren.call(matcherContext, actual, expected);      
      expect(result.message).toMatchSnapshot();
    });

    it('should describe a negated failure appropriately', () => {
      const unexpected = [
        <li data-tis-wife={true} />,
        <li>Lu</li>,
        <div data-awesome={true}>An awesome div!!</div>,
        <span>Lame Span</span>
      ];

      matcherContext.isNot = true;
      const result = toHaveMatchingChildren.call(matcherContext, actual, unexpected);
      expect(result.message).toMatchSnapshot();
    });
  });
});