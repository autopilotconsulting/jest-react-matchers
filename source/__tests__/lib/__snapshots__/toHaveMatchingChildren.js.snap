// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReactMatchers #toHaveMatchingChildren should describe a failure appropriately 1`] = `
"Expected to find children:
  <div data-awesome='true'>
    An awesome div!!
  </div>
  <span>
    Lame Span
  </span>

in:
  <div>
    <li>
      Pippy
    </li>
    <li>
      Hatch
    </li>
    <li data-tis-wife='true'>
      Lu
    </li>
  </div>"
`;

exports[`ReactMatchers #toHaveMatchingChildren should describe a negated failure appropriately 1`] = `
"Expected not to find children:
  <li data-tis-wife='true' />
  <li>
    Lu
  </li>

in:
  <div>
    <li>
      Pippy
    </li>
    <li>
      Hatch
    </li>
    <li data-tis-wife='true'>
      Lu
    </li>
  </div>"
`;

exports[`ReactMatchers #toHaveMatchingChildren should fail if the actual element has no children 1`] = `
"Actual element has no children:
  <div />"
`;
